core/src/main/java/org/apache/brooklyn/core/objs/BasicSpecParameter.java

```diff
@@ -404,7 +431,126 @@ public static void addParameters(AbstractBrooklynObjectSpec<?, ?> spec, List<? e
             spec.parametersAdd(BasicSpecParameter.fromSpec(loader.getManagementContext(), spec));
         }
         if (explicitParams.size() > 0) {
-            spec.parametersAdd(explicitParams);
+            spec.parametersReplace(resolveParameters(explicitParams, spec.getParameters(), spec));
+        }
+    }
+
+    /** merge parameters against other parameters and known and type-inherited config keys */
+    static Collection<SpecParameter<?>> resolveParameters(Collection<? extends SpecParameter<?>> newParams, Collection<? extends SpecParameter<?>> existingReferenceParamsToKeep, AbstractBrooklynObjectSpec<?,?> spec) {
+        Map<String,SpecParameter<?>> existingToKeep = MutableMap.of();
+        if (existingReferenceParamsToKeep!=null) {
+            for (SpecParameter<?> p: existingReferenceParamsToKeep) { 
+                existingToKeep.put(p.getConfigKey().getName(), p);
+            }
+        }
+
+        List<SpecParameter<?>> result = MutableList.<SpecParameter<?>>of();
+        for (SpecParameter<?> p: newParams) {
+            if (p instanceof SpecParameterForInheritance) {
```

Comment:
that would work but i put the extra branch in to show the logic that should be applied; have used your cleanup to extract `existingP` just once, and expanded the placeholder code so it's clearer